### PR TITLE
Cleanup IndexFieldData visibility

### PR DESCRIPTION
**Pull request by s1monw**
Today we expose `IndexFieldDataService` outside of IndexService to do maintainance
or lookup field data in different ways. Yet, we have a streamlined way to access IndexFieldData
via `QueryShardContext` that should encapsulate all access to it. This also ensures that we control
all other functionality like cache clearing etc.

This change also removes the `recycler` option from `ClearIndicesCacheRequest` this option is a no-op and should have been removed long ago.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/admin/indices/cache/clear/ClearIndicesCacheRequest.java [hunk 1]
```diff
@@ -29,10 +31,9 @@
 
     private boolean queryCache = false;
     private boolean fieldDataCache = false;
-    private boolean recycler = false;
     private boolean requestCache = false;
-    private String[] fields = null;
-    
+    private String[] fields = Strings.EMPTY_ARRAY;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/admin/indices/cache/clear/ClearIndicesCacheRequest.java]: that change alone makes me happy

File: core/src/main/java/org/elasticsearch/index/IndexService.java [hunk 2]
```diff
@@ -892,4 +888,37 @@ AsyncTranslogFSync getFsyncTask() { // for tests
...
+            if (fields.length ==  0) {
+                indexCache.clear("api");
+                indexFieldData.clear();
+            } else {
+                // only clear caches relating to the specified fields
+                for (String field : fields) {
+                    indexFieldData.clearField(field);
+                }
+            }
+        }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: I see you did not change that logic but it confuses me a bit as it seems we always clear fileddata even when fieldDataCache is false?

**Review by jpountz — APPROVED** [review]

**Review by jimczi — APPROVED** [review]
LGTM for the code, tests failed to compile ;)

File: core/src/main/java/org/elasticsearch/index/IndexService.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/IndexService.java]: yeah we are on the same page. I will open a followup with a BWC break

**Pull request merged**